File: src/com/mishiranu/instantimage/LoadingDialog.java
package com.mishiranu.instantimage;

import java.util.ArrayList;
import java.util.HashMap;

import android.app.Dialog;
import android.app.DialogFragment;
import android.app.ProgressDialog;
import android.os.Bundle;
import android.util.Pair;

import com.mishiranu.instantimage.async.AsyncManager;
import com.mishiranu.instantimage.async.LoadImageTask;
import com.mishiranu.instantimage.async.LoadQueryTask;

public class LoadingDialog extends DialogFragment implements AsyncManager.Callback
{
	public static final String TAG = LoadingDialog.class.getName();

	private static final String EXTRA_QUERY = "query";
	private static final String EXTRA_IMAGE_URI_STRING = "imageUriString";
	private static final String TASK_LOAD = "load";
	
	public LoadingDialog()
	{
		
	}
	
	public LoadingDialog(String query, String imageUriString, Callback callback)
	{
		Bundle args = new Bundle();
		args.putString(EXTRA_QUERY, query);
		args.putString(EXTRA_IMAGE_URI_STRING, imageUriString);
		setArguments(args);
		setCallback(callback);
	}
	
	public static interface Callback
	{
		public void onQueryLoadingSuccess(ArrayList<ImageItem> imageItems);
		public void onQueryLoadingError(int errorMessageId);
		
		public void onImageLoadingSuccess(int contentId, String mimeType);
		public void onImageLoadingError(int errorMessageId);
	}
	
	private Callback mCallback;
	
	public void setCallback(Callback callback)
	{
		mCallback = callback;
	}
	
	@Override
	public void onDetach()
	{
		super.onDetach();
		mCallback = null;
	}
	
	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState)
	{
		ProgressDialog dialog = new ProgressDialog(getActivity());
		dialog.setMessage(getString(R.string.text_loading));
		dialog.setCanceledOnTouchOutside(false);
		return dialog;
	}
	
	@Override
	public void onActivityCreated(Bundle savedInstanceState)
	{
		super.onActivityCreated(savedInstanceState);
		HashMap<String, Object> extra = new HashMap<>();
		extra.put(EXTRA_QUERY, getArguments().getString(EXTRA_QUERY));
		extra.put(EXTRA_IMAGE_URI_STRING, getArguments().getString(EXTRA_IMAGE_URI_STRING));
		AsyncManager.get(this).startTask(TASK_LOAD, this, extra, false);
	}
	
	@Override
	public Pair<Object, AsyncManager.Holder> onCreateAndExecuteTask(String name, HashMap<String, Object> extra)
	{
		String query = (String) extra.get(EXTRA_QUERY);
		String imageUriString = (String) extra.get(EXTRA_IMAGE_URI_STRING);
		if (query != null)
		{
			LoadQueryTask task = new LoadQueryTask(query);
			task.executeOnExecutor(LoadQueryTask.THREAD_POOL_EXECUTOR);
			return task.getPair();
		}
		else if (imageUriString != null)
		{
			LoadImageTask task = new LoadImageTask(getActivity(), imageUriString);
			task.executeOnExecutor(LoadImageTask.THREAD_POOL_EXECUTOR);
			return task.getPair();
		}
		else return null;
	}
	
	@Override
	public void onFinishTaskExecution(String name, AsyncManager.Holder holder)
	{
		if (getArguments().getString(EXTRA_QUERY) != null)
		{
			int errorMessageId = holder.nextArgument();
			if (errorMessageId != 0) mCallback.onQueryLoadingError(errorMessageId); else
			{
				ArrayList<ImageItem> imageItems = holder.nextArgument();
				if (imageItems == null || imageItems.isEmpty()) mCallback.onQueryLoadingError(R.string.error_not_found);
				else mCallback.onQueryLoadingSuccess(imageItems);
			}
		}
		else if (getArguments().getString(EXTRA_IMAGE_URI_STRING) != null)
		{
			int errorMessageId = holder.nextArgument();
			int contentId = holder.nextArgument();
			String mimeType = holder.nextArgument();
			if (errorMessageId != 0) mCallback.onImageLoadingError(errorMessageId);
			else mCallback.onImageLoadingSuccess(contentId, mimeType);
		}
		dismissAllowingStateLoss();
	}
	
	@Override
	public void onRequestTaskCancel(String name, Object task)
	{
		if (task instanceof LoadQueryTask) ((LoadQueryTask) task).cancel(true);
		else if (task instanceof LoadImageTask) ((LoadImageTask) task).cancel(true);
	}
}

File: src/com/mishiranu/instantimage/ui/LoadingDialog.scala
package com.mishiranu.instantimage.ui

import android.app.{DialogFragment, ProgressDialog}
import android.os.Bundle

import com.mishiranu.instantimage.R
import com.mishiranu.instantimage.async.{LoadImageTask, LoadQueryTask}
import com.mishiranu.instantimage.model.Image

import rx.lang.scala.Subscription

class LoadingDialog extends DialogFragment with RxFragment {
  import LoadingDialog._

  private var callback: Callback = _

  private def this(data: Either[String, Iterable[String]], callback: LoadingDialog.Callback) = {
    this
    val args = new Bundle
    data match {
      case Left(query) =>
        args.putString(EXTRA_QUERY, query)
      case Right(imageUriStrings) =>
        import scala.collection.JavaConverters._
        args.putStringArrayList(EXTRA_IMAGE_URI_STRINGS, new java.util.ArrayList(imageUriStrings.toList.asJava))
    }
    setArguments(args)
    setCallback(callback)
  }

  def this(query: String, callback: LoadingDialog.Callback) = {
    this(Left(query), callback)
  }

  def this(imageUriStrings: Iterable[String], callback: LoadingDialog.Callback) = {
    this(Right(imageUriStrings), callback)
  }

  def setCallback(callback: Callback): Unit = {
    this.callback = callback
  }

  override def onDetach(): Unit = {
    super.onDetach()
    callback = null
  }

  override def onCreateDialog(savedInstanceState: Bundle): ProgressDialog = {
    val dialog = new ProgressDialog(getActivity)
    dialog.setMessage(getString(R.string.text_loading))
    dialog.setCanceledOnTouchOutside(false)
    dialog
  }

  override protected def createRxSubscription: Subscription = {
    val query = getArguments.getString(EXTRA_QUERY)
    val imageUriStrings = getArguments.getStringArrayList(EXTRA_IMAGE_URI_STRINGS)
    if (query != null) {
      LoadQueryTask(query, 0).subscribe(onQueryLoad _)
    } else if (imageUriStrings != null) {
      import scala.collection.JavaConverters._
      LoadImageTask(getActivity, imageUriStrings.asScala.toList).subscribe(onImagesLoad _)
    } else {
      throw new RuntimeException
    }
  }

  private def onQueryLoad(result: LoadQueryTask.Result): Unit = handleRxResult {
    dismiss()
    if (result.errorMessageId != 0) {
      callback.onQueryLoadingError(result.errorMessageId)
    } else if (result.images.isEmpty) {
      callback.onQueryLoadingError(R.string.error_not_found)
    } else {
      callback.onQueryLoadingSuccess(result.images)
    }
  }

  private def onImagesLoad(results: List[LoadImageTask.Result]): Unit = handleRxResult {
    dismiss()
    val successResults = results.filter(_.errorMessageId == 0)
    if (successResults.nonEmpty) {
      callback.onImageLoadingSuccess(successResults.map(r => ImageLoadingSuccessResult(r.contentId, r.mimeType)))
    } else {
      callback.onImageLoadingError(results.map(_.errorMessageId).find(_ != 0).get)
    }
  }
}

object LoadingDialog {
  val TAG: String = classOf[LoadingDialog].getName

  private val EXTRA_QUERY = "query"
  private val EXTRA_IMAGE_URI_STRINGS = "imageUriStrings"

  trait Callback {
    def onQueryLoadingSuccess(images: List[Image]): Unit
    def onQueryLoadingError(errorMessageId: Int): Unit

    def onImageLoadingSuccess(results: List[ImageLoadingSuccessResult]): Unit
    def onImageLoadingError(errorMessageId: Int): Unit
  }

  case class ImageLoadingSuccessResult(contentId: Int, mimeType: String)
}
